### PR TITLE
tracker updates

### PR DESCRIPTION
**Pull request by Maxime-J**
As `fetch` was used to replace `navigator.sendBeacon`, it might be good to handle all the requests with it.
It's more convenient than XHR.

If you're okay, this PR :
- replaces XHR by Fetch
- centralizes the requests (no call to collect api without passing through `collect()`)
- adds a parameter to the collect function, for requests which needs to persist on exit/location change,
like it is currently the case with link events (and it can probably be useful for possible future uses)
- applies the same logic in `sendEvent()` as in other functions

### CODE REVIEW

**Review by rohandebsarkar — APPROVED** [review]

File: tracker/index.js [hunk 1]
```diff
@@ -73,17 +58,22 @@ import { removeTrailingSlash } from '../lib/url';
...
+    if (cache) headers['x-umami-cache'] = cache;
+    let options = {
+      method: 'POST',
+      body: JSON.stringify({type, payload}),
+      headers
+    };
+    if (!persist) {
+      fetch(endpoint, options).then(res => res.text()).then(resText => { cache = resText });
+    } else {
+      options['keepalive'] = true;
```
Comment (rohandebsarkar) [inline tracker/index.js:74]: I think, all the requests should use the `keepalive` option. Previously, the other requests were sent using XHR, therefore those requests were not persisted.

File: tracker/index.js [hunk 2]
```diff
@@ -73,17 +58,22 @@ import { removeTrailingSlash } from '../lib/url';
     return a;
   };
 
-  const collect = (type, payload) => {
+  const collect = (type, payload, persist = false) => {
```
Comment (rohandebsarkar) [inline tracker/index.js:61]:
```suggestion
  const collect = (type, payload, cacheResponse = false) => {
```

File: tracker/index.js [hunk 3]
```diff
@@ -73,17 +58,22 @@ import { removeTrailingSlash } from '../lib/url';
...
-        payload,
-      },
-      res => (cache = res),
-    );
+    const endpoint = `${root}/api/collect`;
+    let headers = { 'Content-Type': 'application/json' };
+    if (cache) headers['x-umami-cache'] = cache;
+    let options = {
+      method: 'POST',
+      body: JSON.stringify({type, payload}),
```
Comment (rohandebsarkar) [inline tracker/index.js:68]:
```suggestion
      body: JSON.stringify({type, payload}),
      keepalive: true,
```

File: tracker/index.js [hunk 4]
```diff
@@ -73,17 +58,22 @@ import { removeTrailingSlash } from '../lib/url';
...
+      headers
+    };
+    if (!persist) {
```
Comment (rohandebsarkar) [inline tracker/index.js:71]:
```suggestion
    if (cacheResponse) {
```

File: tracker/index.js (hunk 1 above)
Comment (rohandebsarkar) [inline tracker/index.js:74]:
```suggestion

```

File: tracker/index.js [hunk 5]
```diff
@@ -105,28 +95,21 @@ import { removeTrailingSlash } from '../lib/url';
...
-      body: data,
-      keepalive: true,
-    });
+  const sendEvent = (event_value, event_type) => {
+    collect(
+      'event',
+      assign(getPayload(), {
+        event_type,
+        event_value,
+      }),
```
Comment (rohandebsarkar) [inline tracker/index.js:110]:
```suggestion
      })
```

File: tracker/index.js [hunk 6]
```diff
@@ -105,28 +95,21 @@ import { removeTrailingSlash } from '../lib/url';
...
+      true
```
Comment (rohandebsarkar) [inline tracker/index.js:111]:
```suggestion

```

File: tracker/index.js [hunk 7]
```diff
@@ -93,7 +83,7 @@ import { removeTrailingSlash } from '../lib/url';
         website: uuid,
         url,
         referrer,
-      }),
+      })
```
Comment (rohandebsarkar) [inline tracker/index.js:86]:
```suggestion
      }),
      true
```

**Review by rohandebsarkar — CHANGES_REQUESTED** [review]
I think this should be the structure for fetch requests.

Whether to cache the response could also be checked by `type === 'pageview'`, but the previous one is probably better if any other responses need to cached in the future.

**Comment by Maxime-J** [comment]
The suggested changes differs a lot from the initial PR, which keeps exactly the same behaviour as now:

Standard requests not resistant to location change/exit, like in XHR, and the answer always cached locally
Specific requests (currently for link events) resistant to location change/exit, like in `navigator.sendBeacon`, without handling the answer cache.

**Comment by rohandebsarkar** [comment]
Yeah, okay. I'll probably work this on another PR if @mikecao is interested.

**Comment by mikecao** [comment]
I'm still undecided on this because the reason we went with XHR is to support older browsers. I haven't read up on the state of fetch support yet. Maybe we can create a separate tracker to support older browsers?

**Comment by rohandebsarkar** [comment]
I understand, but `fetch()` is currently supported by all the major browsers, except Internet Explorer. But Internet Explorer has already been [declared retired on 15 June 2022](https://blogs.windows.com/windowsexperience/2022/06/15/internet-explorer-11-has-retired-and-is-officially-out-of-support-what-you-need-to-know/).

Reference: https://developer.mozilla.org/en-US/docs/Web/API/fetch#browser_compatibility

**Comment by Maxime-J** [comment]
Understood, actually went with fetch because it was introduced in tracker in a previous commit.
And as Rohan said, it's widely supported now.

I also believe that kind of error wouldn't happen (#1246), not sure though.

**Review by mikecao — APPROVED** [review]

**Pull request merged**